File: solutions/day-4/c++/day4.cpp
#include <algorithm>
#include <filesystem>
#include <format>
#include <fstream>
#include <iostream>
#include <iterator>
#include <queue>
#include <ranges>
#include <vector>

constexpr int DELTA_I[8] = {-1, -1, -1, +0, +1, +1, +1, +0},
              DELTA_J[8] = {-1, +0, +1, +1, +1, +0, -1, -1};

struct Grid {
    std::vector<std::string> data;

    Grid(std::vector<std::string> _data) : data(std::move(_data)) {}

    auto operator[](this auto &&self, int i, int j) -> decltype(auto) {
        return self.data[i][j];
    }

    bool inBounds(int i, int j) const {
        return 0 <= i && i < ssize(data) && 0 <= j && j < ssize(data[i]);
    }

    auto positions() const {
        return std::views::cartesian_product(
            std::views::iota(0, static_cast<int>(ssize(data))),
            std::views::iota(0, static_cast<int>(ssize(data[0]))));
    }

    auto neighbors(int i, int j) const {
        return std::views::iota(0, 8) | std::views::transform([=](int k) {
                   return std::pair{i + DELTA_I[k], j + DELTA_J[k]};
               }) |
               std::views::filter([this](auto pos) {
                   return inBounds(pos.first, pos.second);
               });
    }
};
int countAdjacentRolls(int i, int j, const Grid &diagram) {
    int rolls = 0;
    for (auto [ii, jj] : diagram.neighbors(i, j)) {
        rolls += diagram[ii, jj] == '@';
    }
    return rolls;
}

bool accessible(int i, int j, const Grid &diagram) {
    return countAdjacentRolls(i, j, diagram) < 4;
}

bool just_unlocked(int i, int j, const Grid &diagram) {
    return countAdjacentRolls(i, j, diagram) == 3;
}

std::string part1(std::ifstream &inputFile) {
    Grid diagram = {std::views::istream<std::string>(inputFile) |
                    std::ranges::to<std::vector>()};
    auto accessibleRolls =
        std::ranges::count_if(diagram.positions(), [&](auto pos) {
            auto [i, j] = pos;
            return diagram[i, j] == '@' && accessible(i, j, diagram);
        });
    return std::format("There are {} rolls of paper accessible by forklift.",
                       accessibleRolls);
}

std::string part2(std::ifstream &inputFile) {
    Grid diagram = {std::views::istream<std::string>(inputFile) |
                    std::ranges::to<std::vector>()};

    int accessibleRolls = 0;
    std::queue<std::pair<int, int>> toRemove;
    for (auto [i, j] : diagram.positions()) {
        if (diagram[i, j] == '@' && accessible(i, j, diagram))
            toRemove.emplace(i, j);
    }

    while (!toRemove.empty()) {
        accessibleRolls++;
        auto [i, j] = toRemove.front();
        toRemove.pop();
        diagram[i, j] = '.';
        for (auto [ii, jj] : diagram.neighbors(i, j)) {
            if (diagram[ii, jj] == '@' && just_unlocked(ii, jj, diagram))
                toRemove.emplace(ii, jj);
        }
    }

    return std::format(
        "There are {} rolls of paper accessible by forklift with removal.",
        accessibleRolls);
}

int main() {
    namespace fs = std::filesystem;
    fs::path inputFileName = "day4.in";

    if (!fs::exists(inputFileName)) {
        std::cerr << std::format("File {} does not exist.",
                                 inputFileName.string())
                  << std::endl;
        return 1;
    }

    std::ifstream inputFile;
    inputFile.open(inputFileName);

    if (!inputFile) {
        std::cerr << std::format("Couldn't open {}", inputFileName.string())
                  << std::endl;
        return 1;
    }

    std::cout << part1(inputFile) << std::endl;
    inputFile.clear();
    inputFile.seekg(0);
    std::cout << part2(inputFile) << std::endl;
}

File: solutions/day-4/python/day4.py
from pathlib import Path
from itertools import product


class Grid:
    def __init__(self, data):
        self.data = data

    def __getitem__(self, pos):
        i, j = pos
        return self.data[i][j]

    def __setitem__(self, pos, value):
        i, j = pos
        self.data[i][j] = value

    def in_bounds(self, i, j):
        return 0 <= i < len(self.data) and 0 <= j < len(self.data[i])

    def positions(self):
        return product(range(len(self.data)), range(len(self.data[0])))

    def neighbors(self, i, j):
        return ((ii, jj) for ii, jj in product(range(i-1, i+2), range(j-1, j+2)) if (self.in_bounds(ii, jj)) and not (ii == i and jj == j))


def count_adjacent_rolls(i, j, diagram: Grid):
    return sum(1 for (ii, jj) in diagram.neighbors(i, j) if diagram[ii, jj] == '@')


def accessible(i, j, diagram):
    return count_adjacent_rolls(i, j, diagram) < 4


def just_unlocked(i, j, diagram):
    return count_adjacent_rolls(i, j, diagram) == 3


def part1(diagram):
    diagram = Grid(diagram)
    return f"There are {sum(1 for i, j in diagram.positions() if diagram[i,j] == '@' and accessible(i, j, diagram))} rolls of paper accessible by forklift."


def part2(diagram):
    diagram = Grid(diagram)
    queue = [(i, j) for i, j in diagram.positions()
             if diagram[i, j] == '@' and accessible(i, j, diagram)]

    accessibleRolls = 0
    while (len(queue)):
        accessibleRolls += 1
        (i, j) = queue.pop(0)
        diagram[i, j] = '.'
        queue += [(ii, jj)
                  for ii, jj in diagram.neighbors(i, j) if diagram[ii, jj] == '@' and just_unlocked(ii, jj, diagram)]

    return f'There are {accessibleRolls} rolls of paper accessible by forklift with removal.'


file_name = "day4.in"

diagram = [list(s) for s in Path(file_name).read_text().splitlines()]
print(part1(diagram))
print(part2(diagram))
